pg_analytics/src/hooks/query.rs

```diff
@@ -47,15 +51,83 @@ impl Query for *mut pg_sys::PlannedStmt {
         let statement = &ast[0];
 
+        // Convert the AST into a logical plan
         let context_provider = QueryContext::new()?;
         let sql_to_rel = SqlToRel::new(&context_provider);
 
-        Ok(sql_to_rel.statement_to_plan(statement.clone())?)
+        let logical_plan: LogicalPlan;
+
+        // If functions are undefined, then try to find and register the function and then try to get the plan again
+        let re = Regex::new(r"Invalid function '(.+)'")?;
+        loop {
+            match sql_to_rel.statement_to_plan(statement.clone()) {
+                Ok(plan) => {
+                    logical_plan = plan;
+                    break;
+                }
+                Err(err) => match err {
+                    DataFusionError::Plan(err_string) => {
+                        // This regex checks for "Invalid function" in the plan error and
+                        //     otherwise pushes the plan error up, breaking the loop.
+                        let missing_func_name = re
+                            .captures(&err_string)
+                            .ok_or(DataFusionError::Plan(err_string.clone()))?
+                            .get(1)
+                            .ok_or(DataFusionError::Plan(err_string.clone()))?
+                            .as_str();
+
+                        // If we are unable to load the function, we push the error up, breaking the loop
+                        unsafe { loadfunction(missing_func_name)? };
+
+                        // Loop again
+                    }
+                    _ => return Err(ParadeError::DataFusion(err)),
+                },
+            };
+        }
+
+        // Pass UDF name as another argument to UDFs
+        let exprs = logical_plan.expressions();
+        let mut new_exprs = vec![];
+        let mut includes_udf = false;
+        for expr in exprs.iter() {
+            if let Expr::ScalarFunction(ScalarFunction {
+                func_def: ScalarFunctionDefinition::UDF(udf),
+                args,
+            }) = expr
+            {
+                // Pass funcname to udf
+                let mut new_args = args.clone();
+                new_args.splice(
+                    0..0,
+                    vec![Expr::Literal(ScalarValue::Utf8(Some(
+                        udf.name().to_string(),
+                    )))],
+                );
+                new_exprs.push(Expr::ScalarFunction(ScalarFunction {
+                    func_def: ScalarFunctionDefinition::UDF(udf.clone()),
+                    args: new_args,
+                }));
+                includes_udf = true;
+            } else {
+                new_exprs.push(expr.clone());
+            }
+        }
+
+        let mut new_inputs = vec![];
+        for input in logical_plan.inputs().iter() {
+            #[allow(suspicious_double_ref_op)]
+            new_inputs.push(input.clone().clone());
+        }
```

Comment:
Is this double clone intentional? You might be able to get away with:
```rust
let mut new_inputs = logical_plan.inputs()
    .iter()
    .cloned()  // This clones the items in the iterator directly
    .collect::<Vec<_>>();
```

Reply:
Unfortunately double clone is necessary, but using .cloned().cloned() seems to make the warning go away...